File: 3488.closest-equal-element-queries.py
#
# @lc app=leetcode id=3488 lang=python3
#
# [3488] Closest Equal Element Queries
#

# @lc code=start
from collections import defaultdict
from bisect import bisect_left

class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)
        positions = defaultdict(list)
        
        # Group indices by value
        for i, num in enumerate(nums):
            positions[num].append(i)
            
        results = []
        
        for q_idx in queries:
            val = nums[q_idx]
            indices = positions[val]
            k = len(indices)
            
            if k <= 1:
                results.append(-1)
                continue
            
            # Find the position of the current query index in the sorted list of indices
            # Since we know q_idx is in indices, bisect_left will find it efficiently.
            pos_in_list = bisect_left(indices, q_idx)
            
            # Identify left and right neighbors in the circular sense
            # The list 'indices' is sorted.
            # Left neighbor index in 'indices' list:
            left_neighbor_idx = indices[(pos_in_list - 1 + k) % k]
            # Right neighbor index in 'indices' list:
            right_neighbor_idx = indices[(pos_in_list + 1) % k]
            
            # Calculate circular distance function
            def dist(i, j):
                d = abs(i - j)
                return min(d, n - d)
            
            d1 = dist(q_idx, left_neighbor_idx)
            d2 = dist(q_idx, right_neighbor_idx)
            
            results.append(min(d1, d2))
            
        return results
# @lc code=end

File: 3488.closest-equal-element-queries.cpp
#
# @lc app=leetcode id=3488 lang=cpp
#
# [3488] Closest Equal Element Queries
#
# @lc code=start
class Solution {
public:
    vector<int> solveQueries(vector<int>& nums, vector<int>& queries) {
        int n = nums.size();
        unordered_map<int, vector<int>> valueToIndices;
        
        // Group indices by their values
        for (int i = 0; i < n; i++) {
            valueToIndices[nums[i]].push_back(i);
        }
        
        // Precompute minimum distances for each index
        vector<int> minDist(n, -1);
        
        for (auto& [val, indices] : valueToIndices) {
            int k = indices.size();
            if (k == 1) continue; // Only one occurrence, no other index with same value
            
            // Indices are already sorted since we iterate from 0 to n-1
            for (int j = 0; j < k; j++) {
                int curr = indices[j];
                int prev = indices[(j - 1 + k) % k];
                int next = indices[(j + 1) % k];
                
                // Distance to previous neighbor (circular distance)
                int directPrev = abs(curr - prev);
                int distPrev = min(directPrev, n - directPrev);
                
                // Distance to next neighbor (circular distance)
                int directNext = abs(next - curr);
                int distNext = min(directNext, n - directNext);
                
                minDist[curr] = min(distPrev, distNext);
            }
        }
        
        // Answer queries
        vector<int> answer;
        answer.reserve(queries.size());
        for (int q : queries) {
            answer.push_back(minDist[q]);
        }
        
        return answer;
    }
};
# @lc code=end